pkg/blockio/blockio_linux.go

```diff
@@ -0,0 +1,72 @@
+//go:build linux
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package blockio
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/intel/goresctrl/pkg/blockio"
+	runtimespec "github.com/opencontainers/runtime-spec/specs-go"
+
+	"github.com/containerd/containerd/log"
+)
+
+var (
+	enabled   bool
+	enabledMu sync.RWMutex
+)
+
+// IsEnabled checks whether blockio is enabled.
+func IsEnabled() bool {
+	enabledMu.RLock()
+	defer enabledMu.RUnlock()
+
+	return enabled
+}
+
+// SetConfig updates blockio config with a given config path.
+func SetConfig(configFilePath string) error {
```

Comment:
This feels a bit weird to live in a separate location from the config. 

- IIUC, the mutex is to protect concurrency on the `configFilePath`. 
- That path is defined in a [`services/tasks.Config `](https://github.com/containerd/containerd/blob/97480afdac09c947d48f5e3a134db86c78f4bfa6/services/tasks/local.go#L68-L74)
- But updating it now uses a package level mutex


Would it be better to have accessors on `services/tasks.Config`, which can handle concurrency (and in that case concurrency (per config, or globally?)) Perhaps something like https://github.com/moby/locker (or equivalent) if concurrency needs to be "per config file")

Reply:
Hi! 

The issue I run into is that the test-cases use plugin-init to setup memory-level containerd instances and it will be data-race issue when they update the blockio/rdt configs in the same time.

https://github.com/containerd/containerd/blob/97480afdac09c947d48f5e3a134db86c78f4bfa6/integration/build_local_containerd_helper_test.go#L110

> IIUC, the mutex is to protect concurrency on the configFilePath.

I think it is not for the `configFilePath`. For example, the `SetConfigFromFile` will update a package-level var `classBlockIO`.

```go
// https://github.com/intel/goresctrl/blob/cf1db7c9cfe529d718909767ebb3f0686f82d10f/pkg/blockio/blockio.go#L143-L145
// classBlockIO connects user-defined block I/O classes to
// corresponding cgroups blockio controller parameters.
var classBlockIO = map[string]cgroups.BlockIOParameters{}
```

> Perhaps something like https://github.com/moby/locker (or equivalent) if concurrency needs to be "per config file")

Yes. I was thinking about that. The `github.com/intel/goresctrl/pkg/blockio` package doesn't allow us to handle it per config filepath, because it is using package-level var 😞 

That is why I use package-level mutex to ensure that the config-update is safe and `enable`-accessor is routine-safe. Maybe we can file proposal to upstream to use individual object instead of package-level var. And then we use `moby/locker` to protect concurrency on individual config.

WDYT?

cc @klihub @mikebrow

Reply:
/cc @askervin @marquiz

Reply:
@fuweid I try to paraphrase the problem to see if I understood it correctly: Currently both the RDT and blockio bits in `goresctrl` implicitly assume that there is only a single configuration active in any process using `goresctrl`. They also expect that process to take care of any configuration update vs. usage serialization. This is normally true, but the containerd integration test setup where multiple test cases run independently and unaware of other, (currently) violating these assumptions.

Reply:
@klihub 

Yes. If containerd instance is the standalone process, the configure file is only one and `{blockio,rdt}/SetConfigFromFile` is only called once, because there is only one `services/task` plugin instance.

But the containerd repo also provides library mode. With library mode, there maybe several `services/task` plugin instances, for instance, `TestCRIImagePullTimeout` cases. 

https://github.com/containerd/containerd/blob/4c44ec702fccbfdbe1e14382d21a8123fd5774a9/integration/build_local_containerd_helper_test.go#L92-L117

Since there are two cases running in the parallel mode, the following `initBlockIO/initRdt` functions will be called in two goroutines. The `enabled` var will be accessed by two goroutines and test will raise data-race issue.

https://github.com/containerd/containerd/blob/4c44ec702fccbfdbe1e14382d21a8123fd5774a9/services/tasks/blockio_linux.go#L33-L49

For the safety reason,  I use `wrapper` package to provide `update-config` function and it is singleton pattern. 
It doesn't break existing functionality.